azurerm/internal/services/web/resource_arm_app_service_plan.go

```diff
@@ -197,6 +197,14 @@ func resourceArmAppServicePlanCreateUpdate(d *schema.ResourceData, meta interfac
 		return fmt.Errorf("`reserved` has to be set to false when kind is set to `Windows`")
 	}
 
+	if strings.EqualFold(*sku.Tier, "Dynamic") && !strings.EqualFold(kind, "FunctionApp") {
+		return fmt.Errorf("`kind` has to be set to `FunctionApp` when `sku.tier` is `Dynamic`")
+	}
```

Comment:
isn't Dynamic available for Consumption Plan Web Apps too?

Reply:
this PR is just a workaround. `kind` should not be input by user, the backend service will calculate the value for `kind`. But to avoid breaking change, and to make user use it correctly, I added this constraints.

the backend service will always change the kind to `functionapp` if sku tier is `dynamic`. will change kind to `elastic` if sku tier is `ElasticPremium`.